reconstructSPI/iterative_refinement/expectation_maximization.py

```diff
@@ -0,0 +1,524 @@
+                    )
+                    map_3d_f_updated_1 += inserted_slice_3d_r + 1j * inserted_slice_3d_i
+                    counts_3d_updated_1 += count_3d_r + count_3d_i
+
+                for one_slice_idx in range(bayes_factors_2.shape[0]):
+                    xyz = xyz_rotated[one_slice_idx]
+                    inserted_slice_3d_r, count_3d_r = IterativeRefinement.insert_slice(
+                        particle_f_deconv_2.real, xyz, n_pix
+                    )
+                    inserted_slice_3d_i, count_3d_i = IterativeRefinement.insert_slice(
+                        particle_f_deconv_2.imag, xyz, n_pix
+                    )
+                    map_3d_f_updated_2 += inserted_slice_3d_r + 1j * inserted_slice_3d_i
+                    counts_3d_updated_2 += count_3d_r + count_3d_i
+
+            # apply noise model
+            # half_map_1, half_map_2 come from doing the above
+            # independently. Filter by noise estimate (e.g. multiply
+            # both half maps by FSC)
+            fsc_1d = IterativeRefinement.compute_fsc(
+                map_3d_f_updated_1, map_3d_f_updated_2
+            )
+
+            fsc_3d = IterativeRefinement.expand_1d_to_3d(fsc_1d)
+
+            # multiplicative filter on maps with fsc
+            # The FSC is 1D, one number per spherical shells
+            # it can be expanded back to a multiplicative filter of
+            # the same shape as the maps
+            map_3d_f_filtered_1 = map_3d_f_updated_1 * fsc_3d
+            map_3d_f_filtered_2 = map_3d_f_updated_2 * fsc_3d
+
+            # update iteration
+            half_map_3d_f_1 = map_3d_f_filtered_1
+            half_map_3d_f_2 = map_3d_f_filtered_2
+
+        # final map
+        fsc_1d = IterativeRefinement.compute_fsc(half_map_3d_f_1, half_map_3d_f_2)
+        fsc_3d = IterativeRefinement.expand_1d_to_3d(fsc_1d)
+        map_3d_f_final = (half_map_3d_f_1 + half_map_3d_f_2 / 2) * fsc_3d
+        map_3d_r_final = IterativeRefinement.ifft_3d(map_3d_f_final)
+        half_map_3d_r_1 = IterativeRefinement.ifft_3d(half_map_3d_f_1)
+        half_map_3d_r_2 = IterativeRefinement.ifft_3d(half_map_3d_f_2)
+
+        return map_3d_r_final, half_map_3d_r_1, half_map_3d_r_2, fsc_1d
+
+    @staticmethod
+    def split_array(arr):
+        """Split array into two halves along 0th axis.
+
+        Parameters
+        ----------
+        arr : arr
+            Shape (n_particles, ...)
+
+        Returns
+        -------
+        arr1 : arr
+            Shape (n_particles // 2, ...)
+        arr2: arr
+            Shape (n_particles // 2, ...)
+        """
+        if isinstance(arr, (list, tuple)):
+            idx_half = len(arr) // 2
+            arr_1, arr_2 = arr[:idx_half], arr[idx_half:]
+
+            if len(arr_1) != len(arr_2):
+                arr_2 = arr[idx_half : 2 * idx_half]
+        else:
+            idx_half = arr.shape[0] // 2
+            arr_1, arr_2 = arr[:idx_half], arr[idx_half:]
```

Comment:
Hmm, I see a duplicate line for the two cases.

In fact, you don't need to split them up. `len(arr)` will give you `arr.shape[0]`, so perhaps you don't need these different cases and can treat arrays, tupples and lists all the same.

Reply:
Quite amazed I wasn't aware of this haha!